src/database/postgres_test.go

```diff
@@ -0,0 +1,253 @@
+	os.Setenv("DATABASE_URL", "postgres://postgres:postgres@localhost:5432/testdb")
+	defer os.Unsetenv("DATABASE_URL")
+
+	var db *pg.DB
+	for i := 0; i < maxRetries; i++ {
+		db, err = createTestingDatabase()
+		if err == nil || !strings.Contains(err.Error(), "connection reset by peer") {
+			break
+		}
+		t.Logf("Failed to create test database: Connection Reset by peer. Retrying...")
+		time.Sleep(time.Second * 2)
+	}
+	if err != nil {
+		t.Fatalf("Failed to create test database after %d attempts: %v", maxRetries, err)
+	}
+	defer db.Close()
+
+	// Test connection
+	ctx := context.Background()
+	conn, err := establishConnection(ctx)
+
+	assert.NoError(t, err)
+	defer conn.Close(ctx)
+
+	// Test database table initialisation
+	err = initializeDatabaseSchema(ctx, conn)
+	if err != nil {
+		t.Fatalf("Failed to initialise the database schema: %v", err)
+	}
+
+	// Test row insertion
+	err = testInsertRow(ctx, conn)
+	if err != nil {
+		t.Fatalf("Failed to insert row: %v", err)
+	}
+
+	// Test query
+	err = testQueryRow(ctx, conn)
+	if err != nil {
+		t.Fatalf("Failed to query row: %v", err)
+	}
+
+	// Cleanup
+	if err := removeTestingDatabase(); err != nil {
+		t.Fatalf("Failed to remove test database: %v", err)
+	}
+}
+
+func testInsertRow(ctx context.Context, conn *pgx.Conn) error {
+
+	index := &pb.Index{
+		Entries: []*pb.IndexListEntry{
+			{
+				Name:              "Test Entry",
+				URI:               "http://example.com",
+				DataType:          pb.DataType_TEXT,
+				SourceType:        pb.SourceType_WEBPAGE,
+				FirstAddedTime:    timestamppb.Now(),
+				LastRefreshedTime: timestamppb.Now(),
+				Content:           "Test Content",
+				WordOccurrences:   map[string]int32{"test": 1, "content": 1},
+			},
+		},
+	}
+
+	err := insertRows(ctx, conn, index)
+	if err != nil {
+		return fmt.Errorf("failed to insert rows: %v", err)
+	}
+
+	// Delete the test entry
+	_, err = conn.Exec(ctx, "DELETE FROM index_list WHERE name = $1", "Test Entry")
+	if err != nil {
+		return fmt.Errorf("failed to delete test entry: %v", err)
+	}
+
+	return nil
+}
+
+func testQueryRow(ctx context.Context, conn *pgx.Conn) error {
+
+	expectedEntry := &pb.IndexListEntry{
+		Name:              "Test Entry",
+		URI:               "http://example.com",
+		DataType:          pb.DataType_TEXT,
+		SourceType:        pb.SourceType_WEBPAGE,
+		FirstAddedTime:    timestamppb.Now(),
+		LastRefreshedTime: timestamppb.Now(),
+		Content:           "Test Content",
+		WordOccurrences:   map[string]int32{"test": 1, "content": 1},
+	}
+
+	index := &pb.Index{
+		Entries: []*pb.IndexListEntry{expectedEntry},
+	}
+
+	err := insertRows(ctx, conn, index)
+	if err != nil {
+		return fmt.Errorf("failed to insert rows: %v", err)
+	}
+
+	entry, err := queryRow(ctx, conn, "Test Entry")
+	if err != nil {
+		return fmt.Errorf("failed to query row: %v", err)
+	}
+
+	if entry == nil {
+		return fmt.Errorf("entry is nil")
+	}
+
+	if expectedEntry.Name != entry.Name {
```

Comment:
why can you not do a DeepEqual on all these types? It seems overly verbose to test each field like this, and again, it's not flexible in case the proto format changes

Reply:
Good point. Then we need to change it for the subcommands test cases as well.

Reply:
Now using content metadata so we can do deep equal